### ISSUE TITLE
[sled agent] Instance creation will deadlock, stop all future instances from being provisioned, if you give any bad hostnames

### ISSUE THREAD

**Issue by smklein** [oxidecomputer/omicron#4931]
TL;DR:

- If you create a new instance, containing either an invalid hostname, host_domain, or domain_search_list...
- We do some parsing within `Instance::new`, which will return an error.
- This is actually a Bad Thing, because `Instance::new` holds ownership of an `InstanceTicket`, which, while being dropped, grabs a `std::Mutex` to remove itself from the "Map of managed instances"...
- ... which is particularly nasty, because `Instance::new` is called from a context where the "Map of managed instances" is already locked. This results in a deadlock. This also blocks all future provisions, by preventing access to this instance map.

Context:

- Where we first lock the instances in the "instance manager": https://github.com/oxidecomputer/omicron/blob/62547d22e74c4bb134b8d65e373ace793de7529f/sled-agent/src/instance_manager.rs#L240
- Where we create the "instance ticket": https://github.com/oxidecomputer/omicron/blob/62547d22e74c4bb134b8d65e373ace793de7529f/sled-agent/src/instance_manager.rs#L266-L267
- Where we call "instance::new": https://github.com/oxidecomputer/omicron/blob/62547d22e74c4bb134b8d65e373ace793de7529f/sled-agent/src/instance_manager.rs#L288-L295
- Where we can fail within "Instance::new", triggering the deadlock: https://github.com/oxidecomputer/omicron/blob/62547d22e74c4bb134b8d65e373ace793de7529f/sled-agent/src/instance.rs#L728-L747

**Comment by bnaecker** [comment]
Here is some MDB data to prove that this is indeed a deadlock as laid out above.

We took a core of the sled-agent on dogfood, sled 11 (hostname `BRM42220006`), and pointed `mdb` at it. Looking at the stacks we see:

```
BRM42220006 # mdb core.651
Loading modules: [ libumem.so.1 libnvpair.so.1 libc.so.1 ld.so.1 ]
> $G
C++ symbol demangling enabled
> ::stacks
THREAD           STATE    SOBJ                COUNT
2                PARKED   CV                    122
                 libc.so.1`cond_wait_queue+0x55
                 libc.so.1`__cond_wait+0xba
                 libc.so.1`cond_wait+0x2e
                 libc.so.1`pthread_cond_wait+0x15
                 parking_lot::condvar::Condvar::wait_until_internal::hab79cd95889c5f9b+0x27b
                 tokio::runtime::scheduler::multi_thread::park::Parker::park::h83deff8c6c11953a+0x100
                 tokio::runtime::scheduler::multi_thread::worker::Context::park_timeout::h4911e4b52ec44a5e+0x107
                 tokio::runtime::scheduler::multi_thread::worker::Context::run::h92de103105c2abf9+0xfc8
                 tokio::runtime::context::runtime::enter_runtime::h2e88c09b100a744d+0x1a0
                 tokio::runtime::scheduler::multi_thread::worker::run::hda07f908dc95373a+0x4b
                 tokio::runtime::task::core::Core<T,S>::poll::ha503725394676955+0x73
                 tokio::runtime::task::harness::Harness<T,S>::poll::hd2d9b2a75dfbff95+0x97
                 std::sys_common::backtrace::__rust_begin_short_backtrace::h10e4e227e30206c1+0x13d
                 core::ops::function::FnOnce::call_once{{vtable.shim}}::h52cb33c7daa0dcf0+0x78
                 std::sys::unix::thread::Thread::new::thread_start::h1d307f20b84b526d+0x29
                 libc.so.1`_thrp_setup+0x77
                 libc.so.1`_lwp_start

7b               PARKED   MX                      4
                 libc.so.1`mutex_lock_impl+0x113
                 libc.so.1`mutex_lock+0x13
                 _$LT$FuncType$u20$as$u20$dropshot..handler..HttpHandlerFunc$LT$Context$C$$LP$T1$C$T2$RP$$C$ResponseType$GT$$GT$::handle_request::_$u7b$$u7b$closure$u7d$$u7d$::hceac14528ce730cf +0x8be
                 <dropshot::handler::HttpRouteHandler<Context,HandlerType,FuncParams,ResponseType> as dropshot::handler::RouteHandler<Context>>::handle_request::{{closure}}::ha7c9d71453020417+0x3a1
                 dropshot::server::http_request_handle::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h24639d3dbdab461e +0x111
                 tokio::runtime::task::core::Core<T,S>::poll::h0a4254676aa1188d+0x3b
                 tokio::runtime::task::harness::Harness<T,S>::poll::h09f4e8dbc60acb31+0x4d
                 tokio::runtime::scheduler::multi_thread::worker::Context::run_task::h34855a87592a433e+0x280
                 tokio::runtime::scheduler::multi_thread::worker::Context::run::h92de103105c2abf9+0xa4e
                 tokio::runtime::context::runtime::enter_runtime::h2e88c09b100a744d+0x1a0
                 tokio::runtime::scheduler::multi_thread::worker::run::hda07f908dc95373a+0x4b
                 tokio::runtime::task::core::Core<T,S>::poll::ha503725394676955+0x73
                 tokio::runtime::task::harness::Harness<T,S>::poll::hd2d9b2a75dfbff95+0x97
                 std::sys_common::backtrace::__rust_begin_short_backtrace::h10e4e227e30206c1+0x13d
                 core::ops::function::FnOnce::call_once{{vtable.shim}}::h52cb33c7daa0dcf0+0x78
                 std::sys::unix::thread::Thread::new::thread_start::h1d307f20b84b526d+0x29
                 libc.so.1`_thrp_setup+0x77
                 libc.so.1`_lwp_start
<SNIP>
78               PARKED   MX                      1
                 libc.so.1`mutex_lock_impl+0x113
                 libc.so.1`mutex_lock+0x13
                 omicron_sled_agent::instance_manager::InstanceTicket::terminate::h81bc06d79a8cb2d1+0x63
                 omicron_sled_agent::instance::Instance::new::h988e095791e388eb+0x1119
                 _$LT$FuncType$u20$as$u20$dropshot..handler..HttpHandlerFunc$LT$Context$C$$LP$T1$C$T2$RP$$C$ResponseType$GT$$GT$::handle_request::_$u7b$$u7b$closure$u7d$$u7d$::hceac14528ce730cf +0x112d
                 <dropshot::handler::HttpRouteHandler<Context,HandlerType,FuncParams,ResponseType> as dropshot::handler::RouteHandler<Context>>::handle_request::{{closure}}::ha7c9d71453020417+0x3a1
                 dropshot::server::http_request_handle::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h24639d3dbdab461e +0x111
                 tokio::runtime::task::core::Core<T,S>::poll::h0a4254676aa1188d+0x3b
                 tokio::runtime::task::harness::Harness<T,S>::poll::h09f4e8dbc60acb31+0x4d
                 tokio::runtime::scheduler::multi_thread::worker::Context::run_task::h34855a87592a433e+0x280
                 tokio::runtime::scheduler::multi_thread::worker::Context::run::h92de103105c2abf9+0xa4e
                 tokio::runtime::context::runtime::enter_runtime::h2e88c09b100a744d+0x1a0
                 tokio::runtime::scheduler::multi_thread::worker::run::hda07f908dc95373a+0x4b
                 tokio::runtime::task::core::Core<T,S>::poll::ha503725394676955+0x73
                 tokio::runtime::task::harness::Harness<T,S>::poll::hd2d9b2a75dfbff95+0x97
                 std::sys_common::backtrace::__rust_begin_short_backtrace::h10e4e227e30206c1+0x13d
                 core::ops::function::FnOnce::call_once{{vtable.shim}}::h52cb33c7daa0dcf0+0x78
                 std::sys::unix::thread::Thread::new::thread_start::h1d307f20b84b526d+0x29
                 libc.so.1`_thrp_setup+0x77
                 libc.so.1`_lwp_start
```

So we see 4 threads trying to acquire a mutex, inside a Dropshot handler. It's not perfectly clear, but we believe these to be a number of requests to provision an instance to the sled-agent. The last stack is important. It shows one thread also trying to lock a mutex, inside `Instance::new()` and `InstanceTicket::terminate()`.

So it appears that several threads are waiting on this mutex. Who holds it? Let's start by grabbing that thread's stack:

```
> 78::findstack
stack pointer for thread 120 [tokio-runtime-worker]: fffff5ffd05fc240
[ fffff5ffd05fc240 libc.so.1`__lwp_park+0x17() ]
  fffff5ffd05fc290 libc.so.1`mutex_lock_impl+0x113()
  fffff5ffd05fc2b0 libc.so.1`mutex_lock+0x13()
  fffff5ffd05fc310 omicron_sled_agent::instance_manager::InstanceTicket::terminate::h81bc06d79a8cb2d1+0x63()
  fffff5ffd05fcbd0 omicron_sled_agent::instance::Instance::new::h988e095791e388eb+0x1119()
  fffff5ffd05feea0 _$LT$FuncType$u20$as$u20$dropshot..handler..HttpHandlerFunc$LT$Context$C$$LP$T1$C$T2$RP$$C$ResponseType$GT$$GT$::handle_request::_$u7b$$u7b$closure$u7d$$u7d$::hceac14528ce730cf +0x112d()
  fffff5ffd05ff4a0 <dropshot::handler::HttpRouteHandler<Context,HandlerType,FuncParams,ResponseType> as dropshot::handler::RouteHandler<Context>>::handle_request::{{closure}}::ha7c9d71453020417+0x3a1()
  fffff5ffd05ff7e0 dropshot::server::http_request_handle::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h24639d3dbdab461e +0x111()
  fffff5ffd05ff850 tokio::runtime::task::core::Core<T,S>::poll::h0a4254676aa1188d+0x3b()
  fffff5ffd05ffb60 tokio::runtime::task::harness::Harness<T,S>::poll::h09f4e8dbc60acb31+0x4d()
  fffff5ffd05ffbb0 tokio::runtime::scheduler::multi_thread::worker::Context::run_task::h34855a87592a433e+0x280()
  fffff5ffd05ffc60 tokio::runtime::scheduler::multi_thread::worker::Context::run::h92de103105c2abf9+0xa4e()
  fffff5ffd05ffd20 tokio::runtime::context::runtime::enter_runtime::h2e88c09b100a744d+0x1a0()
  fffff5ffd05ffd50 tokio::runtime::scheduler::multi_thread::worker::run::hda07f908dc95373a+0x4b()
  fffff5ffd05ffdc0 tokio::runtime::task::core::Core<T,S>::poll::ha503725394676955+0x73()
  fffff5ffd05ffe20 tokio::runtime::task::harness::Harness<T,S>::poll::hd2d9b2a75dfbff95+0x97()
  fffff5ffd05ffef0 std::sys_common::backtrace::__rust_begin_short_backtrace::h10e4e227e30206c1+0x13d()
  fffff5ffd05fff60 core::ops::function::FnOnce::call_once{{vtable.shim}}::h52cb33c7daa0dcf0+0x78()
  fffff5ffd05fffb0 std::sys::unix::thread::Thread::new::thread_start::h1d307f20b84b526d+0x29()
  fffff5ffd05fffe0 libc.so.1`_thrp_setup+0x77()
  fffff5ffd05ffff0 libc.so.1`_lwp_start()
> fffff5ffd05fc240$C
fffff5ffd05fc290 libc.so.1`mutex_lock_impl+0x113(44c6b50, 0)
fffff5ffd05fc2b0 libc.so.1`mutex_lock+0x13(44c6b50)
fffff5ffd05fc310 omicron_sled_agent::instance_manager::InstanceTicket::terminate::h81bc06d79a8cb2d1+0x63()
fffff5ffd05fcbd0 omicron_sled_agent::instance::Instance::new::h988e095791e388eb+0x1119()
fffff5ffd05feea0 _$LT$FuncType$u20$as$u20$dropshot..handler..HttpHandlerFunc$LT$Context$C$$LP$T1$C$T2$RP$$C$ResponseType$GT$$GT$::handle_request::_$u7b$$u7b$closure$u7d$$u7d$::hceac14528ce730cf +0x112d()
fffff5ffd05ff4a0 <dropshot::handler::HttpRouteHandler<Context,HandlerType,FuncParams,ResponseType> as dropshot::handler::RouteHandler<Context>>::handle_request::{{closure}}::ha7c9d71453020417+0x3a1()
fffff5ffd05ff7e0 dropshot::server::http_request_handle::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h24639d3dbdab461e +0x111()
fffff5ffd05ff850 tokio::runtime::task::core::Core<T,S>::poll::h0a4254676aa1188d+0x3b()
fffff5ffd05ffb60 tokio::runtime::task::harness::Harness<T,S>::poll::h09f4e8dbc60acb31+0x4d()
fffff5ffd05ffbb0 tokio::runtime::scheduler::multi_thread::worker::Context::run_task::h34855a87592a433e+0x280()
fffff5ffd05ffc60 tokio::runtime::scheduler::multi_thread::worker::Context::run::h92de103105c2abf9+0xa4e()
fffff5ffd05ffd20 tokio::runtime::context::runtime::enter_runtime::h2e88c09b100a744d+0x1a0()
fffff5ffd05ffd50 tokio::runtime::scheduler::multi_thread::worker::run::hda07f908dc95373a+0x4b()
fffff5ffd05ffdc0 tokio::runtime::task::core::Core<T,S>::poll::ha503725394676955+0x73()
fffff5ffd05ffe20 tokio::runtime::task::harness::Harness<T,S>::poll::hd2d9b2a75dfbff95+0x97()
fffff5ffd05ffef0 std::sys_common::backtrace::__rust_begin_short_backtrace::h10e4e227e30206c1+0x13d()
fffff5ffd05fff60 core::ops::function::FnOnce::call_once{{vtable.shim}}::h52cb33c7daa0dcf0+0x78()
fffff5ffd05fffb0 std::sys::unix::thread::Thread::new::thread_start::h1d307f20b84b526d+0x29()
fffff5ffd05fffe0 libc.so.1`_thrp_setup+0x77(fffff5ffee5db240)
fffff5ffd05ffff0 libc.so.1`_lwp_start()
```

Ok, so that second output includes the address of the mutex we're contended for. Let's look at that in detail:

```
> 44c6b50::print lwp_mutex_t
{
    flags = {
        flag1 = 0x4
        flag2 = 0
        ceiling = 0
        mbcp_type_un = {
            bcptype = 0
            mtype_rcount = {
                count_type1 = 0
                count_type2 = 0
            }
        }
        magic = 0x4d58
    }
    lock = {
        lock64 = {
            pad = [ 0, 0, 0, 0, 0, 0, 0x1, 0x1 ]
        }
        lock32 = {
            ownerpid = 0
            lockword = 0x1010000
        }
        owner64 = 0x101000000000000
    }
    data = 0xfffff5ffee5db240
}
```

The mutex type includes the LWP of the thread holding it inside its `data` member. So we can figure out who that is:

```
> 44c6b50::print lwp_mutex_t data
data = 0xfffff5ffee5db240
```

Well there you go, that is the thread currently blocked in `mutex_lock()` in the last stack. So we've proven that the thread calling `mutex_lock()` is the current owner of the lock itself.
